### PR TITLE
[1/2] Add support for early stopping during training epoch end

### LINKED ISSUE #7033: Support early stopping during training inside of the early stopping callback
## 🚀 Feature

The ability to early stop during training, as controlled by the callback instead of the training loop.

### Motivation

This simplifies the training loop from manually running these callback hooks if the validation epoch won't be run: https://github.com/PyTorchLightning/pytorch-lightning/blob/5bd3cd5f712b65d38812b27cf957261bb06b83c5/pytorch_lightning/trainer/training_loop.py#L152-L159

Similar to flexibility with checkpointing callbacks, this can eventually enable users to specify separate early stop criteria for both training and validation.

### Pitch

Add an `check_on_train_epoch_end` flag to the callback constructor. See https://github.com/PyTorchLightning/pytorch-lightning/pull/6944/ for a sketch

This flag controls whether we check during training or validation. Because the monitor metric may be in training but not validation, or vice versa, this flag makes the check across these two hooks mutually exclusive.

```
    def on_train_epoch_end(self, trainer, pl_module, outputs) -> None:
        if not self._on_train_epoch_end or self._should_skip_check(trainer):
            return
        self._run_early_stopping_check(trainer)

    def on_validation_end(self, trainer, pl_module):
        if self._on_train_epoch_end or self._should_skip_check(trainer):
            return
        self._run_early_stopping_check(trainer)
```

For parity with existing behavior, by default this flag will be False by default. With this feature enabled, users can specify their callback for early stopping during training as such:

```
stop = EarlyStopping(monitor='abc', min_delta=0.1, patience=0, check_on_train_epoch_end=True)
```

Users could then create multiple such callbacks:
```
train_stop = EarlyStopping(monitor='abc', min_delta=0.1, patience=0, check_on_train_epoch_end=True)
val_stop = EarlyStopping(monitor='val_loss', min_delta=0.5, patience=3)
trainer = Trainer(...., callbacks=[train_stop, val_stop], ...)
trainer.fit(...)
```

### Alternatives

Keep as is

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Fixes #7033
Part 1

In a follow up PR, we can remove https://github.com/PyTorchLightning/pytorch-lightning/blob/5bd3cd5f712b65d38812b27cf957261bb06b83c5/pytorch_lightning/trainer/training_loop.py#L152-L160

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by ananthsub — COMMENTED** [review]
this is blocked on #6921 as the hook ordering is off, meaning we can't access the trainer's logger metrics as expected. the callback for `on_train_epoch_end` runs before `training_epoch_end` in the lightning module

File: pytorch_lightning/callbacks/early_stopping.py [hunk 1]
```diff
@@ -53,6 +53,8 @@ class EarlyStopping(Callback):
             monitored has stopped decreasing and in ``'max'`` mode it will stop when the quantity
             monitored has stopped increasing.
         strict: whether to crash the training if `monitor` is not found in the validation metrics.
+        check_on_train_epoch_end: whether to run early stopping at the end of the training epoch.
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/early_stopping.py:56]: name suggestion: In ModelCheckpoint, we have arguments called `every_n_train_steps` and `every_n_val_epochs`. For consistency, I propose `every_train_epoch`.

**Review by awaelchli — COMMENTED** [review]
> In a follow up PR, we can remove

Have you already tried it? Does it go without complications?

File: pytorch_lightning/callbacks/early_stopping.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/callbacks/early_stopping.py:56]: The checkpoint callback also has the frequency as part of the name. I thought the `on_train_epoch_end` would be clear to map to which callback hooks runs, but open to naming!

**Comment by ananthsub** [comment]
@awaelchli see https://github.com/PyTorchLightning/pytorch-lightning/pull/7069 for the followup

File: tests/callbacks/test_early_stopping.py [hunk 2]
```diff
@@ -333,46 +333,86 @@ def test_early_stopping_mode_options():
...
+                     True,
+                     'ddp_cpu',
+                     2,
+                     marks=RunIf(skip_windows=True)),
+        pytest.param([
+            EarlyStopping(monitor='cba', patience=3, check_on_train_epoch_end=True),
+            EarlyStopping(monitor='abc', check_on_train_epoch_end=True)
+        ],
+                     3,
+                     True,
```
Comment (carmocca) [inline tests/callbacks/test_early_stopping.py:403]: Did `yapf` produce this mess?

File: pytorch_lightning/callbacks/early_stopping.py [hunk 3]
```diff
@@ -53,6 +53,8 @@ class EarlyStopping(Callback):
...
+            If this is ``False``, then the check runs at the end of validation.
```
Comment (carmocca) [inline pytorch_lightning/callbacks/early_stopping.py:57]:
```suggestion
            If this is ``False``, then the check runs at the end of the validation epoch.
```
```suggestion
            If this is ``False``, then the check runs at the end of validation.
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -8,6 +8,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 ## [1.3.0] - 2021-MM-DD
 
 ### Added
+- Added support for the `EarlyStopping` callback to run at the end of the training epoch ([#6944](https://github.com/PyTorchLightning/pytorch-lightning/pull/6944/))
```
Comment (carmocca) [inline CHANGELOG.md:11]:
```suggestion


- Added support for the `EarlyStopping` callback to run at the end of the training epoch ([#6944](https://github.com/PyTorchLightning/pytorch-lightning/pull/6944/))
```
```suggestion
- Added support for the `EarlyStopping` callback to run at the end of the training epoch ([#6944](https://github.com/PyTorchLightning/pytorch-lightning/pull/6944/))
```

File: tests/callbacks/test_early_stopping.py (hunk 2 above)
Comment (ananthsub) [reply tests/callbacks/test_early_stopping.py:403]: Yes

Comment (carmocca) [reply tests/callbacks/test_early_stopping.py]: Pushed a commit to improve it

File: pytorch_lightning/callbacks/early_stopping.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/early_stopping.py:62]: I personally prefer Ananth's option

**Review by carmocca — APPROVED** [review]

File: tests/callbacks/test_early_stopping.py [hunk 5]
```diff
@@ -359,8 +359,13 @@ def validation_epoch_end(self, outputs):
...
 
     # Make sure the trainer stops for the max of all minimum requirements
-    assert trainer.global_step == max(min_steps, by_early_stopping, by_min_epochs), \
-        (trainer.global_step, max(min_steps, by_early_stopping, by_min_epochs), step_freeze, min_steps, min_epochs)
+    assert trainer.global_step == max(min_steps, by_early_stopping, by_min_epochs), (
+        trainer.global_step,
+        max(min_steps, by_early_stopping, by_min_epochs),
+        step_freeze,
+        min_steps,
+        min_epochs,
```
Comment (Borda) [inline tests/callbacks/test_early_stopping.py:369]: can we have some better error formatting at least say what are the variables, this is just bunch of numbers...

File: tests/callbacks/test_early_stopping.py [hunk 6]
```diff
@@ -372,53 +377,122 @@ def test_early_stopping_mode_options():
...
+            None,
+            1,
+        ),
+        pytest.param(
+            [EarlyStopping(monitor='abc'), EarlyStopping(monitor='cba', patience=3)],
+            3,
+            False,
+            'ddp_cpu',
+            2,
+            marks=RunIf(skip_windows=True),
```
Comment (Borda) [inline tests/callbacks/test_early_stopping.py]: why do you skip Windows? the CPU version of DDP shall be supported, right? @awaelchli

File: tests/callbacks/test_early_stopping.py (hunk 5 above)
Comment (awaelchli) [reply tests/callbacks/test_early_stopping.py:369]: @Borda I think this is just the auto formatter and there is actually no change in this PR.

**Review by awaelchli — APPROVED** [review]

File: tests/callbacks/test_early_stopping.py [hunk 7]
```diff
@@ -372,53 +379,97 @@ def test_early_stopping_mode_options():
...
+    def __init__(self, expected_end_epoch: int, early_stop_on_train: bool):
         super().__init__()
         self.expected_end_epoch = expected_end_epoch
+        self.early_stop_on_train = early_stop_on_train
 
-    def validation_epoch_end(self, outputs):
+    def _epoch_end(self) -> None:
         losses = [8, 4, 2, 3, 4, 5, 8, 10]
-        val_loss = losses[self.current_epoch]
-        self.log('abc', torch.tensor(val_loss))
```
Comment (tchaton) [inline tests/callbacks/test_early_stopping.py:382]: Should we log with different keys on val/train for the test ?

**Review by tchaton — APPROVED** [review]
LGTM !

File: tests/callbacks/test_early_stopping.py [hunk 8]
```diff
@@ -372,53 +379,116 @@ def test_early_stopping_mode_options():
...
     def on_train_end(self) -> None:
         assert self.trainer.current_epoch == self.expected_end_epoch, 'Early Stopping Failed'
 
 
 @pytest.mark.parametrize(
-    "callbacks, expected_stop_epoch, accelerator, num_processes",
+    "callbacks, expected_stop_epoch, check_on_train_epoch_end, accelerator, num_processes",
     [
-        ([EarlyStopping(monitor='abc'), EarlyStopping(monitor='cba', patience=3)], 3, None, 1),
+        ([EarlyStopping(monitor='abc'), EarlyStopping(monitor='cba', patience=3)], 3, False, None, 1),
```
Comment (Borda) [inline tests/callbacks/test_early_stopping.py:410]: @carmocca id we simplify each EarlyStop we can fit to single line :P

Comment (carmocca) [reply tests/callbacks/test_early_stopping.py:410]: What do you mean?

File: pytorch_lightning/callbacks/early_stopping.py [hunk 9]
```diff
@@ -107,6 +110,7 @@ def __init__(
         self.divergence_threshold = divergence_threshold
         self.wait_count = 0
         self.stopped_epoch = 0
+        self._check_on_train_epoch_end = check_on_train_epoch_end
```
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/callbacks/early_stopping.py:113]: preferring to keep it private out of backwards compatibility concerns. i'm not sure if external callers need to inspect this, but we can open this up in a later PR if it's needed

File: tests/callbacks/test_early_stopping.py (hunk 5 above)
Comment (ananthsub) [reply tests/callbacks/test_early_stopping.py:369]: too late to bring up using `black` as the formatter? 😅

Comment (carmocca) [reply tests/callbacks/test_early_stopping.py:369]: @Borda 😈

Comment (Borda) [reply tests/callbacks/test_early_stopping.py:369]: @ananthsub you can use @carmocca way... format `black` (conservative) and the `yapf` (liberal) :P

File: tests/callbacks/test_early_stopping.py [hunk 10]
```diff
@@ -372,53 +379,69 @@ def test_early_stopping_mode_options():
...
+        ([EarlyStopping('abc'), EarlyStopping('cba', patience=3)], 3, False, None, 1),
+        ([EarlyStopping('cba', patience=3), EarlyStopping('abc')], 3, False, None, 1),
+        pytest.param([EarlyStopping('abc'), EarlyStopping('cba', patience=3)], 3, False, 'ddp_cpu', 2, **_NO_WIN),
+        pytest.param([EarlyStopping('cba', patience=3), EarlyStopping('abc')], 3, False, 'ddp_cpu', 2, **_NO_WIN),
+        ([EarlyStopping('abc', **_ES_CHECK), EarlyStopping('cba', **_ES_CHECK_P3)], 3, True, None, 1),
+        ([EarlyStopping('cba', **_ES_CHECK_P3), EarlyStopping('abc', **_ES_CHECK)], 3, True, None, 1),
+        pytest.param([EarlyStopping('abc', **_ES_CHECK),
+                      EarlyStopping('cba', **_ES_CHECK_P3)], 3, True, 'ddp_cpu', 2, **_NO_WIN),
+        pytest.param([EarlyStopping('cba', **_ES_CHECK_P3),
+                      EarlyStopping('abc', **_ES_CHECK)], 3, True, 'ddp_cpu', 2, **_NO_WIN),
```
Comment (Borda) [inline tests/callbacks/test_early_stopping.py:424]: I mean this @carmocca from https://github.com/PyTorchLightning/pytorch-lightning/pull/6944#discussion_r621314431
